storage/allocator.go

```diff
@@ -104,25 +103,24 @@ type Allocator struct {
 	storePool *StorePool
 	randGen   *rand.Rand
 	options   RebalancingOptions
+	balancer  balancer
 }
 
 // MakeAllocator creates a new allocator using the specified StorePool.
 func MakeAllocator(storePool *StorePool, options RebalancingOptions) Allocator {
-	return Allocator{
+	var randSource rand.Source
+	if options.Deterministic {
+		randSource = rand.NewSource(777)
+	} else {
+		randSource = rand.NewSource(rand.Int63())
+	}
+	a := Allocator{
 		storePool: storePool,
-		randGen:   rand.New(rand.NewSource(rand.Int63())),
+		randGen:   rand.New(randSource),
 		options:   options,
 	}
-}
-
-// getUsedNodes returns a set of node IDs which are already being used
-// to store replicas.
-func getUsedNodes(existing []roachpb.ReplicaDescriptor) map[roachpb.NodeID]struct{} {
-	usedNodes := map[roachpb.NodeID]struct{}{}
-	for _, replica := range existing {
-		usedNodes[replica.NodeID] = struct{}{}
-	}
-	return usedNodes
+	a.balancer = defaultBalancer{a}
```

Comment:
Being separate is an improvement, but the allocator does need to own the balancer.  The backreference was only so that the balancer could use the Allocator's random generator (needed for deterministic tests).

I have changed it so that the rand.Rand is now passed directly to the defaultBalancer, and defaultBalancer no longer has a reference back to the Allocator.